data_structures/binary_search_tree2.cpp

```diff
@@ -0,0 +1,539 @@
+     *
+     * @param new_value The value to insert into the BST.
+     * @return true If the insertion was successful.
+     * @return false Otherwise.
+     */
+    bool insert(T new_value) {
+        bool result = insert(root_, new_value);
+        if (result) {
+            size_++;
+        }
+        return result;
+    }
+
+    /**
+     * @brief Remove a specified value from the BST.
+     *
+     * @param rm_value The value to remove.
+     * @return true If the removal was successful.
+     * @return false Otherwise.
+     */
+    bool remove(T rm_value) {
+        bool result = remove(root_, root_, rm_value);
+        if (result) {
+            size_--;
+        }
+        return result;
+    }
+
+    /**
+     * @brief Check if a value is in the BST.
+     *
+     * @param value The value to find.
+     * @return true If value is in the BST.
+     * @return false Otherwise.
+     */
+    bool contains(T value) { return contains(root_, value); }
+
+    /**
+     * @brief Find the smallest value in the BST.
+     *
+     * @param ret_value Variable to hold the minimum value.
+     * @return true If minimum value was successfully found.
+     * @return false Otherwise.
+     */
+    bool find_min(T& ret_value) { return find_min(root_, ret_value); }
+
+    /**
+     * @brief Find the largest value in the BST.
+     *
+     * @param ret_value Variable to hold the maximum value.
+     * @return true If maximum value was successfully found.
+     * @return false Otherwise.
+     */
+    bool find_max(T& ret_value) { return find_max(root_, ret_value); }
+
+    /**
+     * @brief Get the number of values in the BST.
+     *
+     * @return std::size_t Number of values in the BST.
+     */
+    std::size_t size() { return size_; }
+
+    /**
+     * @brief Get all values of the BST in in-order order.
+     *
+     * @return std::vector<T> List of values, sorted in in-order order.
+     */
+    std::vector<T> get_elements_inorder() {
+        std::vector<T> result;
+        traverse_inorder([&](T node_value) { result.push_back(node_value); },
+                         root_);
+        return result;
+    }
+
+    /**
+     * @brief Get all values of the BST in pre-order order.
+     *
+     * @return std::vector<T> List of values, sorted in pre-order order.
+     */
+    std::vector<T> get_elements_preorder() {
+        std::vector<T> result;
+        traverse_preorder([&](T node_value) { result.push_back(node_value); },
+                          root_);
+        return result;
+    }
+
+    /**
+     * @brief Get all values of the BST in post-order order.
+     *
+     * @return std::vector<T> List of values, sorted in post-order order.
+     */
+    std::vector<T> get_elements_postorder() {
+        std::vector<T> result;
+        traverse_postorder([&](T node_value) { result.push_back(node_value); },
+                           root_);
+        return result;
+    }
+};
+
+/**
+ * @brief Function for testing insert().
+ */
```

Comment:
```suggestion
  * @returns `void`
  */
```
we have to add this stub due to a bug in `doxygen`